### PR TITLE
Add gtk3 and xdgdesktopportal in apprun hook

### PR DESCRIPTION
**Pull request by realnc**
Gtk-based desktops use either the gtk2 or gtk3 plugin. All others use the
XDG DP one (provides native dialog windows.)

Depends on https://github.com/linuxdeploy/linuxdeploy-plugin-qt/pull/62.

### CODE REVIEW

File: src/deployment.h [hunk 1]
```diff
@@ -92,17 +99,23 @@ inline bool createAppRunHook(appdir::AppDir &appDir) {
...
-        << "case \"${XDG_CURRENT_DESKTOP}\" in" << std::endl
-        << "    *GNOME*|*gnome*|*XFCE*)" << std::endl
-        << "        export QT_QPA_PLATFORMTHEME=gtk2" << std::endl
-        << "        ;;" << std::endl
-        << "esac" << std::endl;
-
+    const bool haveGtk3 = bf::exists(themesPath / "libqgtk3.so");
+
+    ofs <<
+        "# generated by linuxdeploy-plugin-qt\n"
```
Comment (TheAssassin) [inline src/deployment.h]: Please use `std::endl` instead of `\n` to remain consistent with all other file generations.

File: src/deployment.h [hunk 2]
```diff
@@ -70,6 +70,13 @@ inline bool createQtConf(appdir::AppDir &appDir) {
 }
 
 inline bool createAppRunHook(appdir::AppDir &appDir) {
+    auto themesPath = appDir.path() / "usr/plugins/platformthemes";
+
+    if (!bf::exists(themesPath) || bf::is_empty(themesPath)) {
```
Comment (TheAssassin) [inline src/deployment.h]: I don't think we need this check, it should be safe to add the script in any case. Why did you add it? Just curious.

File: src/deployment.h [hunk 3]
```diff
@@ -92,17 +99,23 @@ inline bool createAppRunHook(appdir::AppDir &appDir) {
...
-        << "esac" << std::endl;
-
+    const bool haveGtk3 = bf::exists(themesPath / "libqgtk3.so");
+
+    ofs <<
+        "# generated by linuxdeploy-plugin-qt\n"
+        "\n"
+        "# try to make Qt apps more \"native looking\" on Gtk-based desktops, if possible\n"
+        "# on other desktops, use the XDG Desktop Portal plugin instead to get native dialogs\n"
+        "# see https://github.com/AppImage/AppImageKit/issues/977#issue-462374883\n"
```
Comment (TheAssassin) [inline src/deployment.h]: Let's add this PR's URL to the list.

**Comment by TheAssassin** [comment]
Regarding the `haveGtk3` check: doesn't this assume that as soon as a Gtk+3 platform theme lib is found, the app is forced to use Gtk+3 even if the desktop uses Gtk+2?

I am not entirely sure, but I think for Gtk+2 desktops like Xfce `libqgtk3.so` might not work. Have you tested that yet?

File: src/deployment.h (hunk 1 above)
Comment (realnc) [reply src/deployment.h]: Are you sure? There is no flush needed here. (`std::endl` is a shorthand for `std::flush << '\n'`). The stream will get flushed at scope exit.

File: src/deployment.h (hunk 2 above)
Comment (realnc) [reply src/deployment.h]: Well, having the hook is useless without the theme plugins, and since it's an easy check it seemed like a good idea. I'll remove it.

**Comment by realnc** [comment]
> Regarding the `haveGtk3` check: doesn't this assume that as soon as a Gtk+3 platform theme lib is found, the app is forced to use Gtk+3 even if the desktop uses Gtk+2?
> 
> I am not entirely sure, but I think for Gtk+2 desktops like Xfce `libqgtk3.so` might not work. Have you tested that yet?

If the gtk3 platform theme exists, the gtk2 one doesn't. I don't think any Qt version had both.

**Comment by TheAssassin** [comment]
Yeah, seems so. Just tried a few in a container. Thanks for clarifying.

File: src/deployment.h [hunk 4]
```diff
@@ -92,14 +94,20 @@ inline bool createAppRunHook(appdir::AppDir &appDir) {
         return false;
     }
 
+    const bool haveGtk3 = bf::exists(themesPath / "libqgtk3.so");
```
Comment (TheAssassin) [inline src/deployment.h]: Can we add a comment above this variable explaining that it's an either/or, i.e., there's either libqgtk3 or libqgtk2, please?

**Comment by TheAssassin** [comment]
Thanks a lot for helping improve this tool!

**Comment by Shatur95** [comment]
@TheAssassin, @realnc, GTK3 platformtheme currently has [a bug](https://bugreports.qt.io/browse/QTBUG-72506) that results to ugly look. Maybe it worth to not bundle `libqgtk3` for now?

At the moment I have to remove this library from the system to make my application look native.
